### PR TITLE
Allow Bears to use raw files

### PR DESCRIPTION
**Pull request by yukiisbored**
Processing allows raw files if the Bear state that it uses raw
files. Bear that uses binary files can't be mixed with Bear that uses
text files. If this is enabled, bears are incharge of doing the file
handling (opening it, closing it, reading it, etc).

Closes https://github.com/coala/coala/issues/3529

### Checklist

- [x] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [x] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [x] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [x] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [x] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [x] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

### CODE REVIEW

**Comment by sils** [comment]
hey, I don't understand why we need this. Bears can use raw files right now and just ignore whatever coala gives them and that works reasonably well. Also the case of mixing bears that use raw files with native ones works. Your change will make that *not work* and users won't like that. Also it adds more code and more stuff I need to add to my bear and I don't see any benefit. Is there one that I miss?

**Comment by jayvdb** [comment]
@yukiisbored , please rebase this.

**Comment by jayvdb** [comment]
> Bears can use raw files right now and just ignore whatever coala gives them and that works reasonably well.

@sils, They can only do that if `files` has dummy values, and then all the 'modified files' functionality is broken.

> Also it adds more code and more stuff I need to add to my bear and I don't see any benefit.

No, this provides a default value, which only needs to be modified if the bear is happy to accept raw files.

This is blocking three binary bears:

- [ ] https://github.com/coala/coala-bears/pull/1288 - ImageDimensionBear
- [ ] https://github.com/coala/coala-bears/pull/1302 - ImageCompressionBear
- [ ] https://github.com/coala/coala-bears/pull/1543 - MP3 Bear

I need the first two for WMID.

File: coalib/processes/Processing.py [hunk 1]
```diff
@@ -365,8 +370,18 @@ def instantiate_processes(section,
     # only on changed files or not. Global bears require all the files
     complete_filename_list = filename_list
 
+    # Checks if this section uses raw files or not
+    bears = chain(local_bear_list, global_bear_list)
+    use_raw_files = set(bear.USE_RAW_FILES for bear in bears)
+    if len(use_raw_files) > 1:
+        log_printer.err("Bears that uses raw files can't be mixed with "
+                        'Bears that uses text files. Skipping this '
+                        'section.')
```
Comment (sils) [inline coalib/processes/Processing.py]: can we make them mixable? If they're mixed just load a raw and a normal file dict into memory or so?

Comment (jayvdb) [inline coalib/processes/Processing.py]: That introduces lots of edge cases, and so far we have no sensible use case for that.  All raw bears we have in the pipeline will be loading files that cant be loaded as text file.

File: tests/processes/ProcessingTest.py [hunk 2]
```diff
@@ -494,47 +529,6 @@ def test_yield_ignore_ranges(self):
         self.assertEqual(source_range.end.line, 1)
         self.assertEqual(source_range.end.column, 14)
 
-    def test_loaded_bears_with_error_result(self):
```
Comment (yukiisbored) [inline tests/processes/ProcessingTest.py]: Whoops, accidentally removed this during rebase.

File: tests/processes/ProcessingTest.py [hunk 3]
```diff
@@ -36,8 +37,7 @@
 p=subprocess.Popen([sys.executable,
                   "-c",
                   "import time; time.sleep(0.1)"]);
-pgid = p.pid if platform.system() == "Windows" else os.getpgid(p.pid);
-print(p.pid, pgid)
+pgid = p.pid if platform.system() == "Windows" else os.getpgid(p.pid);print(p.pid, pgid)
```
Comment (yukiisbored) [inline tests/processes/ProcessingTest.py]: okay, I have no idea how this happened.

File: tests/processes/ProcessingTest.py [hunk 4]
```diff
@@ -155,6 +157,63 @@ def test_empty_run(self):
...
+                         ", message='test msg', aspect=NoneType\\) at "
+                         '0x[0-9a-fA-F]+>'.format(hex(local_result.id)))
+        self.assertRegex(repr(global_result),
+                         "<Result object\\(id={}, origin='GlobalTestRawBear', "
+                         'affected_code=\\(.*start=.*file=.*section_executor_'
+                         'test_files.*line=None.*end=.*\\), severity=NORMAL, c'
+                         "onfidence=100, message='test message', "
+                         'aspect=NoneType\\'
+                         ') at 0x[0-9a-fA-F]+>'.format(hex(global_result.id)))
+
```
Comment (yukiisbored) [inline tests/processes/ProcessingTest.py]: nu-uh.

File: coalib/processes/Processing.py [hunk 5]
```diff
@@ -385,8 +396,16 @@ def instantiate_processes(section,
     # the whole project is accessible to every bear. However, local bears are
     # run only for the changed files if caching is enabled.
 
+    if len(use_raw_files) > 1:
+        log_printer.err("Bears that uses raw files can't be mixed with "
+                        'Bears that uses text files. Skipping this '
+                        'section.')
+        mixed = True
+    use_raw_files = use_raw_files.pop() if len(use_raw_files) != 0 else False
```
Comment (Mixih) [inline coalib/processes/Processing.py]: Isn't this going to always make use_raw_file False since the set {True, False} will pop False first because of current hash ordering? The pop also makes me a bit uncomfortable since it spec'ed as non-deterministic in the python docs.

File: coalib/processes/Processing.py [hunk 6]
```diff
@@ -385,8 +396,16 @@ def instantiate_processes(section,
     # the whole project is accessible to every bear. However, local bears are
     # run only for the changed files if caching is enabled.
 
+    if len(use_raw_files) > 1:
+        log_printer.err("Bears that uses raw files can't be mixed with "
+                        'Bears that uses text files. Skipping this '
+                        'section.')
```
Comment (sils) [inline coalib/processes/Processing.py]: I think it's crucial here to propose a solution to the user. E.g. "Move the following bears into an own section: ..."

The user doesn't care if a bear uses raw files or not and he also doesn't know so this message needs to be very specific about how to solve this IMO

File: coalib/processes/Processing.py (hunk 5 above)
Comment (yukiisbored) [inline coalib/processes/Processing.py]: Actually, it expects it to be only one object inside the Set. If there's more than one, this section will be skipped anyway since it's mixing bears that uses raw files and the ones that doesn't.

the `if len(use_raw_files) != 0` is added because there are some tests that executes this function without giving a bear.

File: coalib/processes/Processing.py [hunk 7]
```diff
@@ -337,9 +342,10 @@ def instantiate_processes(section,
                           global_bear_list,
                           job_count,
                           cache,
-                          log_printer,
```
Comment (jayvdb) [inline coalib/processes/Processing.py]: why are you removing this?

File: coalib/processes/Processing.py [hunk 8]
```diff
@@ -337,9 +342,10 @@ def instantiate_processes(section,
...
+                          use_raw_files,
```
Comment (jayvdb) [inline coalib/processes/Processing.py]: Never add parameters in the middle.  That is a breaking change, which causes errors like the following.  Move it to the end, and add a default value.

```
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/coalib/coala_main.py", line 122, in run_coala
    debug=debug or args and args.debug)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/coalib/processes/Processing.py", line 763, in execute_section
    debug=debug)
TypeError: instantiate_processes() got multiple values for argument 'console_printer'
```

File: coalib/processes/Processing.py (hunk 7 above)
Comment (yukiisbored) [inline coalib/processes/Processing.py]: Whoops, accidentally deleted it during rebase.

File: tests/processes/section_executor_test_files/ProcessingGlobalTestRawFileBear.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+from coalib.bears.GlobalBear import GlobalBear
+from coalib.results.Result import Result
+
+
+class ProcessingGlobalTestRawFileBear(GlobalBear):  # pragma: no cover
```
Comment (jayvdb) [inline tests/processes/section_executor_test_files/ProcessingGlobalTestRawFileBear.py]: why no cover?
Add a comment above explaining it.
I am trying very hard (https://github.com/coala/coala/pull/4339) to remove "pragma: no cover", so only very specific exceptions are allowed.

File: tests/processes/section_executor_test_files/ProcessingLocalTestRawFileBear.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import time
+
+from coalib.bears.LocalBear import LocalBear
+from coalib.results.Result import Result
+
+
+class ProcessingLocalTestRawFileBear(LocalBear):  # pragma: no cover
```
Comment (jayvdb) [inline tests/processes/section_executor_test_files/ProcessingLocalTestRawFileBear.py]: per other comment.

File: tests/processes/ProcessingTest.py [hunk 11]
```diff
@@ -155,6 +158,66 @@ def test_empty_run(self):
...
+                                  None,
+                                  self.log_printer,
+                                  console_printer=self.console_printer)
+        self.assertTrue(results[0])
+
+        local_results = self.result_queue.get(timeout=0)
+        global_results = self.result_queue.get(timeout=0)
+        self.assertTrue(self.result_queue.empty())
+
+        self.assertEqual(len(local_results), 1)
```
Comment (yukiisbored) [inline tests/processes/ProcessingTest.py]: Should I put it into another commit? Because I'm using the same way as the other one.

File: tests/processes/ProcessingTest.py [hunk 12]
```diff
@@ -329,6 +387,25 @@ def test_get_file_dict_non_existent_file(self):
                        'an unknown error.'),
                       self.log_printer.log_queue.get().message)
 
+    def test_get_file_dict_allow_raw_file(self):
+        file_dict = get_file_dict([self.unreadable_path], self.log_printer,
+                                  True)
+        self.assertNotEqual(file_dict, {})
+        self.assertEqual(file_dict[self.unreadable_path], None)
+        self.assertNotIn(("Failed to read file '{}'. It seems to contain "
```
Comment (jayvdb) [inline tests/processes/ProcessingTest.py]: An idea that I just had over at https://github.com/coala/coala/pull/4270#discussion_r121594930 , ...

It would be more useful to assert that there are no warnings (or errors) in the message queue.

And as part of that, instead of `self.log_printer.log_queue.get().message`, fetch all messages into an array, so that if another message is emitted somewhere in new code before the target message, it doesnt cause this test to silently fail.

**Comment by jayvdb** [comment]
ack 095aeb3

**Comment by jayvdb** [comment]
@rultor merge
